### PR TITLE
LUCENE-10121: More skipping in WANDScorer.

### PR DESCRIPTION
**Pull request by jpountz**
#11158

### CODE REVIEW

**Comment by jpountz** [comment]
I had to add a loop, which makes the patch bigger than it should be due to indentation changes. It's best to review the patch with `?w=1`: https://github.com/apache/lucene/pull/319/files?w=1.

File: lucene/core/src/java/org/apache/lucene/search/WANDScorer.java [hunk 1]
```diff
@@ -385,42 +387,69 @@ private void advanceTail() throws IOException {
...
+      // a competitive hit.
+      while (tailSize > 0 && tailMaxScore >= minCompetitiveScore) {
+        DisiWrapper w = popTail();
+        w.doc = w.iterator.advance(target);
+        head.add(w);
+      }
+
+      // Make sure the head is now beyond the target
+      DisiWrapper top = head.top();
+      if (top != null) {
```
Comment (gsmiller) [inline lucene/core/src/java/org/apache/lucene/search/WANDScorer.java:444]: Maybe check `if (head.size() != 0)` instead to be consistent with the checks earlier in this method?

**Review by gsmiller — COMMENTED** [review]
This `updateMaxScores` method is getting pretty challenging to reason about (for me at least), but I think this looks good. Nice find!

**Comment by jpountz** [comment]
I agree with you that this change is making the code harder to reason about. I'll look into ways to make the logic easier.

**Comment by gsmiller** [comment]
Thanks @jpountz. It might go a long way to just add some comments about the invariants that have to be maintained when updating the score, etc.

**Comment by jpountz** [comment]
Agreed. Maybe another way of fixing it would consist of preserving more significant bits in the scaled values. I'll look into that as well.

**Comment by mikemccand** [comment]
@jpountz is this still relevant?  There have been lots of optos to `WANDScorer` lately... maybe this is already essentially done?

**Comment by jpountz** [comment]
It's still relevant but I'm not comfortable with the fact that it's a bit fragile. I'll close for now and think more about it.

**Pull request closed without merging**